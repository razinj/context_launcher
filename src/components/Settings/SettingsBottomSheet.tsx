// React
import React, { useContext } from 'react'
// React Native
import { StyleSheet, View } from 'react-native'
// Components
import SettingsHeader from './SettingsHeader'
import ToggleSettings from './shared/ToggleSettings'
import RecentAppsSettings from './sections/RecentAppsSettings'
import PinnedAppsSettings from './sections/PinnedAppsSettings'
import FavoriteAppsSettings from './sections/FavoriteAppsSettings'
import AdvancedSettings from './sections/AdvancedSettings'
// Contexts
import GlobalContext from '../../contexts/GlobalContext'
// BottomSheet
import { BottomSheetModal, BottomSheetModalProvider } from '@gorhom/bottom-sheet'

const SettingsBottomSheet = () => {
  const { settingsBottomSheetRef } = useContext(GlobalContext)

  return (
    <BottomSheetModalProvider>
      <BottomSheetModal
        ref={settingsBottomSheetRef}
        snapPoints={['25%', '50%', '75%', '95%']}
        style={styles.bottomSheetModal}>
        <View style={styles.settingsWrapper} testID='settings-bottom-sheet-wrapper'>
          <SettingsHeader />

          <ToggleSettings title={'Recent Apps'}>
            <RecentAppsSettings />
          </ToggleSettings>

          <ToggleSettings title={'Pinned Apps'}>
            <PinnedAppsSettings />
          </ToggleSettings>

          <ToggleSettings title={'Favorite Apps'}>
            <FavoriteAppsSettings />
          </ToggleSettings>

          <ToggleSettings title={'Advanced Settings'}>
            <AdvancedSettings />
          </ToggleSettings>
        </View>
      </BottomSheetModal>
    </BottomSheetModalProvider>
  )
}

const styles = StyleSheet.create({
  bottomSheetModal: {
    marginHorizontal: 5,
  },
  settingsWrapper: {
    paddingHorizontal: 10,
  },
})

export default SettingsBottomSheet
