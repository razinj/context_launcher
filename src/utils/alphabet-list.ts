import { AppDetails } from '../models/app-details'
import { AppLetterIndex } from '../models/list-letter-index'

type LetterMap = {
  [key: string]: number
}

const alphabet = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
]

/**
 * @returns Indexed alphabet letters in an object where the key is the alphabet and the value is the index
 */
export const getLettersMap = (): LetterMap => {
  const letterMap: LetterMap = {}

  alphabet.forEach((letter: string, index: number) => (letterMap[letter] = index + 1))

  return letterMap
}

/**
 * @param value Any string value
 * @returns A letter if the first chareceter of the value is an alphabet, otherwise it returns a '#'
 */
export const getFirstLetter = (value: string): string => {
  const firstChar = value.trim().substring(0, 1).toUpperCase()
  const isValidLetter = getLettersMap()[firstChar]

  if (isValidLetter) return firstChar

  return '#'
}

export const getAppsLetterIndex = (apps: AppDetails[]): AppLetterIndex[] => {
  const appsLetterIndex: AppLetterIndex[] = []
  const treatedLetters: string[] = []

  apps.forEach(({ name }: AppDetails, index: number) => {
    const letter = getFirstLetter(name)

    if (treatedLetters.includes(letter)) return

    appsLetterIndex.push({ letter, index })
    treatedLetters.push(letter)
  })

  return appsLetterIndex
}
