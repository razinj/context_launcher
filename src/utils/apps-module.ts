// Native modules
import AppsModule from '../native-modules/AppsModule'

export const launchApp = (packageName: string) => AppsModule.launchApplication(packageName)

export const showAppDetails = (packageName: string) => AppsModule.showApplicationDetails(packageName)

export const requestAppUninstall = (packageName: string) => AppsModule.requestApplicationUninstall(packageName)
