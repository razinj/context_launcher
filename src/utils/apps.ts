// Models
import { AppDetails, AppDetailsWithIcon } from '../models/app-details'
import { PinnedApp } from '../models/pinned-app'

export const getAppsByLabel = (apps: AppDetails[], query: string): AppDetails[] => {
  return apps.filter((app: AppDetails) => app.label.match(new RegExp(query, 'gi')))
}

export const getPinnedAppByName = (pinnedApps: PinnedApp[], appDetails: AppDetailsWithIcon): PinnedApp => {
  const pinnedApp = pinnedApps.find(({ name }: PinnedApp) => name === appDetails.name)

  if (pinnedApp) return pinnedApp

  return { ...appDetails, isTemporary: false, isPermanent: false }
}
