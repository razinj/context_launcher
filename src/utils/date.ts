export const getTimeFromDate = (stringDate?: string): string => {
  if (!stringDate) return 'unset'

  const date = new Date(stringDate)

  const singleDigitHours = `${date.getHours()}`.length == 1
  const singleDigitMinutes = `${date.getMinutes()}`.length == 1

  const hours = singleDigitHours ? `0${date.getHours()}` : date.getHours()
  const minutes = singleDigitMinutes ? `0${date.getMinutes()}` : date.getMinutes()

  return `${hours}:${minutes}`
}

export const getDateFromStringWithCurrentDateValue = (stringDate?: string): Date | undefined => {
  if (!stringDate) return

  const currentDate = new Date()
  const fullDate = new Date(stringDate)
  return new Date(
    currentDate.getFullYear(),
    currentDate.getMonth(),
    currentDate.getDate(),
    fullDate.getHours(),
    fullDate.getMinutes(),
    0,
    0
  )
}

export const stripDateFromSeconds = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), 0, 0)
}
