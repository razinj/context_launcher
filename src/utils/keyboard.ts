import { Keyboard } from 'react-native'

export const dismissKeyboard = () => Keyboard.dismiss()
