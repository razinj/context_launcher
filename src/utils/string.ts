export const truncateString = (value: string | undefined, maxLength: number): string | undefined => {
  return value && value.length > maxLength ? `${value.substring(0, maxLength)}...` : value
}
