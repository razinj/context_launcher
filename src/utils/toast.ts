import { ToastAndroid } from 'react-native'

export const displayToast = (message: string, duration: number = ToastAndroid.SHORT) => {
  ToastAndroid.show(message, duration)
}
