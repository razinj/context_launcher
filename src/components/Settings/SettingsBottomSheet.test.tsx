import React from 'react'
import { fireEvent, screen } from '@testing-library/react-native'
import { renderWithProvider } from '../../../utils/test/utils'
import { CONTEXT_LAUNCHER_APP_ID } from '../../../utils/test/data'
import SettingsBottomSheet from './SettingsBottomSheet'
import * as AppsModuleUtils from '../../utils/apps-module'

describe('<SettingsBottomSheet /> Tests', () => {
  beforeAll(() => {
    jest.spyOn(AppsModuleUtils, 'showAppDetails')
  })

  it('should render correctly and match snapshot', () => {
    renderWithProvider(<SettingsBottomSheet />)

    expect(screen.toJSON()).toMatchSnapshot()
    // Wrappers
    expect(screen.getByTestId('settings-bottom-sheet-wrapper')).toBeOnTheScreen()
    expect(screen.getByTestId('settings-bottom-sheet-header-wrapper')).toBeOnTheScreen()
    // Elements
    expect(screen.getByTestId('settings-bottom-sheet-header-app-info-button')).toBeOnTheScreen()
  })

  it('should call native module function to open app info when pressed', () => {
    renderWithProvider(<SettingsBottomSheet />)

    const appInfoButton = screen.getByTestId('settings-bottom-sheet-header-app-info-button')

    expect(appInfoButton).toBeOnTheScreen()

    fireEvent.press(appInfoButton)

    expect(AppsModuleUtils.showAppDetails).toBeCalledWith(CONTEXT_LAUNCHER_APP_ID)
    // TODO: Make sure 'settingsBottomSheetRef?.current?.dismiss()' is called
  })
})
