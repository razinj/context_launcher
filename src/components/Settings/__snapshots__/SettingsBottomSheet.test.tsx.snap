// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsBottomSheet /> Tests should render correctly and match snapshot 1`] = `
<View>
  <View
    snapPoints={
      [
        "25%",
        "50%",
        "75%",
        "95%",
      ]
    }
    style={
      {
        "marginHorizontal": 5,
      }
    }
  >
    <View
      style={
        {
          "paddingHorizontal": 10,
        }
      }
      testID="settings-bottom-sheet-wrapper"
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingBottom": 20,
          }
        }
        testID="settings-bottom-sheet-header-wrapper"
      >
        <Text
          style={
            {
              "color": "#808080",
              "fontSize": 20,
              "fontWeight": "400",
            }
          }
        >
          Context Settings
          <Text
            style={
              {
                "color": "#ccc",
                "fontSize": 12,
              }
            }
          >
              v
            1.4.8
             (
            12
            )
          </Text>
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="settings-bottom-sheet-header-app-info-button"
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#808080",
                  "fontSize": 34,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰋽
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "borderRadius": 5,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
          testID="settings-bottom-toggle-advanced-settings-button"
        >
          <View
            style={
              [
                {
                  "padding": 10,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#808080",
                  },
                  [],
                ]
              }
            >
              Recent Apps
            </Text>
          </View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#808080",
                  "fontSize": 30,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰅀
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "borderRadius": 5,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
          testID="settings-bottom-toggle-advanced-settings-button"
        >
          <View
            style={
              [
                {
                  "padding": 10,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#808080",
                  },
                  [],
                ]
              }
            >
              Pinned Apps
            </Text>
          </View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#808080",
                  "fontSize": 30,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰅀
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "borderRadius": 5,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
          testID="settings-bottom-toggle-advanced-settings-button"
        >
          <View
            style={
              [
                {
                  "padding": 10,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#808080",
                  },
                  [],
                ]
              }
            >
              Favorite Apps
            </Text>
          </View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#808080",
                  "fontSize": 30,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰅀
          </Text>
        </View>
      </View>
      <View
        style={
          {
            "borderRadius": 5,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
          testID="settings-bottom-toggle-advanced-settings-button"
        >
          <View
            style={
              [
                {
                  "padding": 10,
                },
                undefined,
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#808080",
                  },
                  [],
                ]
              }
            >
              Advanced Settings
            </Text>
          </View>
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "#808080",
                  "fontSize": 30,
                },
                undefined,
                {
                  "fontFamily": "Material Design Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            󰅀
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
